kotsadm/pkg/kurl/join_cert.go

```diff
@@ -0,0 +1,199 @@
+	}
+
+	return buf.Bytes(), nil
+}
+
+func parseCertGenOutput(logs []byte) (string, error) {
+	// Output looks like this:
+	//
+	// I0806 21:27:41.711156      41 version.go:251] remote version is much newer: v1.18.6; falling back to: stable-1.17
+	// W0806 21:27:41.826204      41 validation.go:28] Cannot validate kube-proxy config - no validator is available
+	// W0806 21:27:41.826231      41 validation.go:28] Cannot validate kubelet config - no validator is available
+	// [upload-certs] Storing the certificates in Secret "kubeadm-certs" in the "kube-system" Namespace
+	// [upload-certs] Using certificate key:
+	// 7cf895f013c8977a24d3603c1802b78af146124d4c3223696b888a53352f4026
+
+	scanner := bufio.NewScanner(bytes.NewReader(logs))
+	foundKeyText := false
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+		if foundKeyText {
+			return line, nil
+		}
+		if strings.Contains(line, "Using certificate key") {
+			foundKeyText = true
+		}
+	}
+
+	return "", errors.New("key not found in output")
+}
+
+func getPodSpec(clientset kubernetes.Interface, namespace string) (*corev1.Pod, error) {
+	existingDeployment, err := clientset.AppsV1().Deployments(namespace).Get(context.TODO(), "kotsadm", metav1.GetOptions{})
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to get existing deployment")
+	}
+
+	apiContainerIndex := -1
+	for i, container := range existingDeployment.Spec.Template.Spec.Containers {
+		if container.Name == "kotsadm" {
+			apiContainerIndex = i
+			break
+		}
+	}
+	if apiContainerIndex == -1 {
+		return nil, errors.New("kotsadm container not found")
+	}
+
+	securityContext := corev1.PodSecurityContext{
+		RunAsUser: util.IntPointer(0),
+	}
+
+	configVolumeType := corev1.HostPathDirectory
+	binVolumeType := corev1.HostPathFile
+	name := fmt.Sprintf("kurl-join-cert-%d", time.Now().Unix())
+	pod := &corev1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "Pod",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: existingDeployment.Namespace,
+			Labels:    existingDeployment.Labels,
+		},
+		Spec: corev1.PodSpec{
+			SecurityContext: &securityContext,
+			NodeSelector: map[string]string{
+				"node-role.kubernetes.io/master": "",
+			},
+			RestartPolicy:    corev1.RestartPolicyNever,
+			ImagePullSecrets: existingDeployment.Spec.Template.Spec.ImagePullSecrets,
+			Volumes: []corev1.Volume{
+				{
+					Name: "config",
+					VolumeSource: corev1.VolumeSource{
+						HostPath: &corev1.HostPathVolumeSource{
+							Path: "/etc/kubernetes/",
+							Type: &configVolumeType,
+						},
+					},
+				},
+				{
+					Name: "kubeadm",
+					VolumeSource: corev1.VolumeSource{
+						HostPath: &corev1.HostPathVolumeSource{
+							Path: "/usr/bin/kubeadm",
+							Type: &binVolumeType,
+						},
+					},
+				},
+			},
+			Containers: []corev1.Container{
+				{
+					Image:           existingDeployment.Spec.Template.Spec.Containers[apiContainerIndex].Image,
```

Comment:
:thinking: the effort to get the container image is appreciated but also seems hard and potentially fragile. 

I am assuming this is needed for airgap installs? Can't we just look for the kurl config map / image pull secret with a label or some other methods to look it up and construct it? Maybe not, this isn't available in the downward api either.

No objections here, but it feels tricky and brittle.